src/substrate/substrate.handler.ts

```diff
@@ -127,53 +128,56 @@ export class SubstrateService implements OnModuleInit {
     }
   }
 
-  async listenToEvents() {
-    await this.api.query.system.events(async (events) => {
-      try {
-        const currentBlock        = await this.api.rpc.chain.getBlock();
-        const currentBlockNumber  = currentBlock.block.header.number.toNumber();
-        const blockHash           = await this.api.rpc.chain.getBlockHash(currentBlockNumber);
-  
-        const blockMetaData: BlockMetaData = {
-          blockNumber: currentBlockNumber,
-          blockHash: blockHash.toString()
-        }
+  async eventFromBlock(blockNumber: number, blockHash: string | Uint8Array) {
+    const apiAt = await this.api.at(blockHash);
+
+    const allEventsFromBlock = await apiAt.query.system.events();
+
+    const events = allEventsFromBlock.filter(({ phase }) => phase.isApplyExtrinsic);
+
+    const blockMetaData: BlockMetaData = {
+      blockNumber: blockNumber,
+      blockHash: blockHash.toString()
+    }
   
-        for (let i = 0; i < events.length; i++) {
-          const { event } = events[i];
-          await this.handleEvent(blockMetaData, event);
-        }
-      } catch (err) {
-        this.logger.log(`Handling listen to event catch : ${err.name}, ${err.message}, ${err.stack}`);
-      }
-    }).then(_unsub => {
-      this.event = _unsub;
-    }).catch(err => {
-      this.logger.log(`Event listener catch error ${err.name}, ${err.message}, ${err.stack}`);
-    });
+    for (let i = 0; i < events.length; i++) {
+      const { event } = events[i];
+      await this.handleEvent(blockMetaData, event);
+    }
   }
 
   async listenToNewBlock() {
     await this.api.rpc.chain.subscribeNewHeads(async (header: Header) => {
       try {
         const blockNumber = header.number.toNumber();
+        const blockHash   = await this.api.rpc.chain.getBlockHash(blockNumber);
 
-        // check if env is development
-        if (process.env.NODE_ENV === 'development') {
-          const lastBlockNumber = await this.queryBus.execute(
+        if (this.lastBlockNumber == 0) {
+          this.lastBlockNumber = await this.queryBus.execute(
             new GetLastSubstrateBlockQuery(),
           );
+        }
 
+        // check if env is development
+        if (process.env.NODE_ENV === 'development') {
```

Comment:
Create a `ProcessEnv` module so testing this will be a lot easier. Using `process.env` directly prevents people from mocking it.

You can refer here: https://github.com/debionetwork/debio-backend/tree/main/src/common/process-env